File: sandbox-runner/patterns/non-retryable-errors/typescript/workflows.ts
import { log, proxyActivities } from "@temporalio/workflow";

import type * as activities from "./activities";

const { processOrder } = proxyActivities<typeof activities>({
  startToCloseTimeout: "5 seconds",
  retry: {
    maximumAttempts: 5,
    initialInterval: "1s",
    backoffCoefficient: 1,
  },
});

export async function orderWorkflow(orderId: string): Promise<string> {
  try {
    return await processOrder(orderId);
  } catch (err) {
    log.error("Order processing failed", { orderId, error: err });
    return `Order ${orderId} failed: ${(err as Error).message}`;
  }
}

File: sandbox-runner/patterns/non-retryable-errors/python/workflows.py
from datetime import timedelta

from temporalio import workflow
from temporalio.common import RetryPolicy
from temporalio.exceptions import ActivityError

with workflow.unsafe.imports_passed_through():
    import activities


@workflow.defn
class OrderWorkflow:
    @workflow.run
    async def run(self, order_id: str) -> str:
        try:
            return await workflow.execute_activity(
                activities.process_order,
                order_id,
                start_to_close_timeout=timedelta(seconds=5),
                retry_policy=RetryPolicy(
                    maximum_attempts=5,
                    initial_interval=timedelta(seconds=1),
                    backoff_coefficient=1.0,
                ),
            )
        except ActivityError as e:
            return f"Order {order_id} failed: {e.cause}"
